### PR TITLE
Describable will give a warning on multiple implementations of a symbol

### PR DESCRIPTION
**Pull request by dgarzon**
This PR addresses https://github.com/jenkinsci/configuration-as-code-plugin/issues/658 and https://github.com/jenkinsci/configuration-as-code-plugin/issues/657 on the CasC side of things. Currently, if there are duplicate symbols both the `configure` and `getImplementors` methods throw ugly exceptions. Ideally, they should not throw and instead log a warning to the user to report the duplicates to maintainers.

I may have gone a bit overboard with the refactoring, but I think it is valuable nonetheless. The code now is easier to reason about as I have moved code to functions and given them clear names. Also tried to DRY this as much as possible.

Behavior:
1. `getImplementors`: if a duplicate key is found, we discard it, and log `Found multiple implementations for symbol = fooBar: [foo.bar.FooBar, foo.baz.FooBar]. Please report to plugin maintainer.`
2. `configure`: if a duplicate key is found, we log `Found multiple implementations for symbol = fooBar: [foo.bar.FooBar, foo.baz.FooBar]. Please report to plugin maintainer.` and configure the first instance returned by `getDescriptors()` which is the same behavior we had before this refactoring.

### CODE REVIEW

**Comment by casz** [comment]
Build failure: https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fconfiguration-as-code-plugin/detail/PR-674/1/pipeline#log-245

**Comment by casz** [comment]
:muscle: I have too much power @timja

**Comment by dgarzon** [comment]
Thank you @casz and @timja I completely forgot about this PR :)

**Comment by casz** [comment]
@BlueAndi this should hopefully fix your duplicate schema issue and together with #777 everything should be.

Honestly me too @dgarzon :laughing:

**Comment by casz** [comment]
CI is having a field day :cry:

**Comment by casz** [comment]
oh @dgarzon apparently it won't actually compile: https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fconfiguration-as-code-plugin/detail/PR-674/23/pipeline#step-59-log-545

**Comment by BlueAndi** [comment]
@casz Thanks for pointing this out. I hope I can check this tomorrow in the office.

**Comment by casz** [comment]
@dgarzon this breaks binary compatibility.

**Comment by casz** [comment]
@dgarzon and @timja please take a look :smile:

**Comment by casz** [comment]
Moved my extra work to a separate PR, it started to become more work than I expected :smile:

**Comment by dgarzon** [comment]
@casz will take a look tomorrow, I feel responsible to get this fixed :)

**Comment by casz** [comment]
@dgarzon :eyes: :see_no_evil:

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/configurators/HeteroDescribableConfigurator.java [hunk 1]
```diff
@@ -57,117 +64,168 @@ public HeteroDescribableConfigurator(Class<T> clazz) {
...
     }
 
-    public Map<String, Class> getImplementors() {
-        final Class api = getImplementedAPI();
-        final List<Descriptor> descriptors = Jenkins.getInstance().getDescriptorList(target);
-        return descriptors.stream()
-            .collect(Collectors.toMap(
-                    d -> DescribableAttribute.getSymbols(d, api, target).get(0),
-                    d -> d.getKlass().toJavaClass()));
+    @SuppressWarnings("unused")
```
Comment (timja) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/configurators/HeteroDescribableConfigurator.java]: do you need this suppression given its used in tests?

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/configurators/HeteroDescribableConfigurator.java [hunk 2]
```diff
@@ -57,117 +64,168 @@ public HeteroDescribableConfigurator(Class<T> clazz) {
...
+
+    private String preferredSymbol(Descriptor<?> descriptor) {
+        return DescribableAttribute.getPreferredSymbol(descriptor, target, target);
+    }
+
+    private Boolean lookupPlugin(String name) {
+        return Option.of(Jenkins.getInstance().getPlugin(name)).isDefined();
+    }
+
+    private HashMap<String, Descriptor<T>> handleDuplicateSymbols(HashMap<String, Descriptor<T>> r, Tuple2<String, Descriptor<T>> t) {
```
Comment (timja) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/configurators/HeteroDescribableConfigurator.java]: it's not very clear what the parameters are here, what are r and t?

also do we need to be using HashMap or can we use Map?

File: support/src/main/java/io/jenkins/plugins/casc/support/jobdsl/SeedJobConfigurator.java [hunk 3]
```diff
@@ -80,7 +80,7 @@ public Class getTarget() {
 
     @Nonnull
     @Override
-    public List<Configurator> getConfigurators(ConfigurationContext context) {
+    public List<Configurator<GeneratedItems[]>> getConfigurators(ConfigurationContext context) {
```
Comment (timja) [inline support/src/main/java/io/jenkins/plugins/casc/support/jobdsl/SeedJobConfigurator.java]: looks unrelated?

Comment (casz) [reply support/src/main/java/io/jenkins/plugins/casc/support/jobdsl/SeedJobConfigurator.java]: No, it is very much related. The Configurator class was changed to actually need the type defined and not done through reflection. Hence the binary incompatibility fix.

**Comment by casz** [comment]
@dgarzon :pray:

**Comment by timja** [comment]
@casz this ready to go?

**Comment by casz** [comment]
It is, I just want @dgarzon to have look at it again with my additions :sweat_smile:

**Comment by dgarzon** [comment]
Thanks @casz for taking care of this! Looks good on my end :) Been super busy at work, hopefully I will free up a bit and be back here to contribute more often!

**Comment by casz** [comment]
:raised_hands:

**Comment by BlueAndi** [comment]
If the descriptor list of the describable attribute is empty, another problem raise up in the [schema](https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/plugin/src/main/resources/io/jenkins/plugins/casc/DescribableAttribute/schema.jelly), because of the "oneOf"
```
...
  "schema" : {
    "oneOf": [
  <j:forEach items="${app.getDescriptorList(it.type)}" var="d" varStatus="st">
...
```
which expects at least one item.

**Comment by casz** [comment]
@BlueAndi wrap it in a app.getDescriptorList(it.type).size if statement?

**Comment by BlueAndi** [comment]
@casz Sounds good, I will try. I saw that here the same problem appears: [schema](https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/plugin/src/main/resources/io/jenkins/plugins/casc/impl/configurators/HeteroDescribableConfigurator/schema.jelly)
Can I use it.implementors.entrySet().size here?
```
...
  "${it.target.name}": {
    "type": "object",
    "oneOf" : [
  <j:forEach items="${it.implementors.entrySet()}" var="impl" varStatus="st">
...
```

**Comment by BlueAndi** [comment]
Previous mentioned problem fixed in db44a1e162207aab258645d475b8ae52910f30e7

Regarding the
```
...
  "schema" : {
    "oneOf": [
  <j:forEach items="${app.getDescriptorList(it.type)}" var="d" varStatus="st">
...
```
I don't get the point about the whole construction, as I think "schema" misses a dollar sign as well.
